File: Java/src/main/java/org/codecop/dependencies/a/Money.java
package org.codecop.dependencies.a;

import java.math.BigDecimal;
import java.util.Objects;

public class Money {

    public static final Money ONE_THOUSAND = new Money(1000);
    public static final Money ONE_HOUNDRED = new Money(100);

    private final BigDecimal value;

    public Money(int value) {
        this(new BigDecimal(value));
    }

    public Money(BigDecimal value) {
        this.value = value;
    }

    public Money reduceBy(int p) {
        return new Money(value.multiply(new BigDecimal(100 - p)).divide(new BigDecimal(100)));
    }

    public boolean moreThan(Money other) {
        return this.value.compareTo(other.value) > 0;
    }

    @Override
    public boolean equals(Object other) {
        if (other == null || getClass() != other.getClass()) {
            return false;
        }
        Money money = (Money) other;
        return Objects.equals(value, money.value);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(value);
    }

}

File: PHP/src/A/Money.php
<?php

namespace Org\Codecop\Dependencies\A;

use Brick\Math\BigDecimal;
// see https://github.com/brick/math

class Money {

    public static function ONE_THOUSAND():Money {
        return new Money(1000);
    }
    public static function ONE_HUNDRED():Money {
        return new Money(100);
    }

    private $value;

    public function __construct($value) {
        if (is_int($value)) {
            $this->value = BigDecimal::of($value);
        } elseif ($value instanceof BigDecimal) {
            $this->value = $value;
        }
    }

    public function reduceBy(int $p): Money {
        $percentage = BigDecimal::of(100)->minus($p);
        return new Money($this->value->multipliedBy($percentage)->exactlyDividedBy(100));
    }

    public function moreThan(Money $other): bool {
        return $this->value->compareTo($other->value) > 0;
    }
}
